### PR TITLE
Fix for Python 3.7

### PR DESCRIPTION
**Pull request by Dormouse759**
This pull request attempts to solve incompatibilities with Python 3.7.
Backwards compatibility shouldn't be broken.

### CODE REVIEW

**Comment by Traceur759** [comment]
The coverage problem is due to tests not being run against 3.7.

File: eventlet/green/ssl.py [hunk 1]
```diff
@@ -57,11 +58,41 @@ def __init__(self, sock, keyfile=None, certfile=None,
...
+
+                self = _original_wrap_socket(
+                    sock=sock.fd,
+                    keyfile=keyfile,
+                    certfile=certfile,
+                    server_side=server_side,
+                    cert_reqs=cert_reqs,
+                    ssl_version=ssl_version,
+                    ca_certs=ca_certs,
+                    do_handshake_on_connect=do_handshake_on_connect and six.PY2,
```
Comment (hroncok) [inline eventlet/green/ssl.py]: the py2 part here is redundant given that the entire section is in if > 3.7 block

File: eventlet/green/ssl.py [hunk 2]
```diff
@@ -57,11 +58,40 @@ def __init__(self, sock, keyfile=None, certfile=None,
...
-            ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
+        if sys.version_info >= (3, 7):
+            # Monkey-patch the sslsocket so our modified self gets
+            # injected into its _create method.
+            def fake_new(self, cls, *args, **kwargs):
+                return self
+
+            orig_new = _original_sslsocket.__new__
+            try:
+                _original_sslsocket.__new__ = fake_new.__get__(self, GreenSSLSocket)
```
Comment (jstasiak) [inline eventlet/green/ssl.py]: Do I understand correctly that this patches/unpatches the original SSLSocket class? If that's the case there's gonna be a race condition - if another thread uses SSLSocket after we patch it but before we unpatch it'll break the application if there's no global monkey patching. Unfortunately I have no competing solution ready right now.

Comment (Traceur759) [reply eventlet/green/ssl.py]: Yes, you understand it correctly.
That was also a problem that crossed my mind, but I couldn't think of better way since __init__ is no longer a thing for SSLSocket and the _create class method calls __new__ with it's own class as first argument. Because of this, there is no way of overriding the behavior of _create method.
TL;DR: There is no way of making _create use subclass' instance.

**Comment by temoto** [comment]
@Traceur759 tests are run against 3.7 and it produces recursion in traceback. Could you please look at it?

https://travis-ci.org/eventlet/eventlet/jobs/405772034
https://travis-ci.org/eventlet/eventlet/jobs/405772035

**Comment by Traceur759** [comment]
The problem with 3.7 is that the buildbot is using outdated version of 3.7.
In that version, the wrap_socket calls SSlSocket, which calls __init__ and vice versa.
In the release version, it calls SSLContext instead, and that doesn't cause this recursion.

**Comment by temoto** [comment]
@Traceur759 can you fix recursion with some condition check or code refactor and enable 3.7 in travis? Maybe they already have `- python: 3.7` without -dev suffix.

**Comment by hroncok** [comment]
See for 3.7 final on Travis: https://github.com/pyvec/elsa/commit/e5a832a1afbb460eb5ea7fc6d05a3f680669bfcd

**Comment by Traceur759** [comment]
@temoto Fixing that would mean supporting a dev version, I don't see a point in that.
Using released 3.7 is itself a fix for the recursion.

File: .travis.yml [hunk 3]
```diff
@@ -22,9 +22,9 @@ matrix:
     - {python: 3.6, env: TOXENV=py36-poll}
     - {python: 3.6, env: TOXENV=py36-selects}
 
-    - {python: 3.7-dev, env: TOXENV=py37-epolls}
-    - {python: 3.7-dev, env: TOXENV=py37-poll}
-    - {python: 3.7-dev, env: TOXENV=py37-selects}
+    - {python: 3.7, env: TOXENV=py37-epolls}
+    - {python: 3.7, env: TOXENV=py37-poll}
+    - {python: 3.7, env: TOXENV=py37-selects}
```
Comment (hroncok) [inline .travis.yml]: Note that `sudo: required` and `dist: xenial` was required last time I've tried.

**Comment by temoto** [comment]
This is what was failing, basically, deprecation warning from third-party package.

```
venv-37/lib/python3.7/site-packages/dns/namedict.py:33: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
  class NameDict(collections.MutableMapping):
```

**Comment by temoto** [comment]
@jstasiak @hroncok do you have more to say or approve or reject?

**Comment by hroncok** [comment]
I don't like the race condition, but this is still better than current state.

**Comment by jstasiak** [comment]
I'd rather have a code that obviously doesn't work rather than something that can result in subtle issues *but* I think I have an idea how to solve this so we don't have to choose - it'll take few days though as I'm a bit busy at the moment.

**Comment by onovy** [comment]
hi, any progress wit Python 3.7 please? This is blocker for Debian Buster, where we want to have Python 3.7 as default and Eventlet is really important lib for for example OpenStack. Thanks.

**Comment by Traceur759** [comment]
@onovy 
I think everyone is currently waiting on @jstasiak to come up with the solution or to at least provide some info.
In the meantime, it should be okay to backport this patch as a temporary solution.

**Comment by nat-goodspeed** [comment]
@jstasiak, is there any way I could be helpful to your better solution?

File: eventlet/green/ssl.py (hunk 2 above)
Comment (kashirin-alex) [reply eventlet/green/ssl.py]: if to stay with a fake_new, may be it is to assign it much earlier at the initial patch import,
https://github.com/eventlet/eventlet/blob/master/eventlet/patcher.py#L416

**Comment by jstasiak** [comment]
@nat-goodspeed sorry, I went down a bit of a rabbit hole testing different approaches to this, I'll have something ready today, thanks for the offer though.

**Comment by jstasiak** [comment]
Sorry, I need to give up, it turned out to be much trickier than I expected. I have no better solution at the moment and I won't have time to work on one for a week or so.

**Comment by nat-goodspeed** [comment]
What would be the least invasive upstream change that would permit a robust solution on the eventlet side?

File: eventlet/green/ssl.py (hunk 2 above)
Comment (Traceur759) [reply eventlet/green/ssl.py]: Wouldn't that just mean higher probability of breaking other apps trying to use SSLsocket?
The less instructions between assignment and reassignment, the lower chance for the race-condition going wrong.

**Comment by temoto** [comment]
@jstasiak @nat-goodspeed ping

**Comment by nat-goodspeed** [comment]
What I was trying to get at with my previous question is: What could we submit as a pull request against the official Python library that would permit a robust eventlet-side solution to this problem? Presumably the smaller/simpler/more-obvious such an upstream change is, the more likely it is to be accepted.
But I haven't dug into this problem as much as others have, which is why I pose it as a question rather than suggesting such an upstream pull request myself.

**Comment by temoto** [comment]
IMHO, as Python3.7 is spreading, it is important to merge non-ideal solution and work on robust later.

Use case 1, old project, say OpenStack: customer says we must support Python3.7, - here's patch, few things don't work, something something and eventlet, - not again! are you kidding... the pure Python lib. OK postpone 3.7 later. - This is angry user.
Use case 2, new project: hey we could use this lib for network with code that doesn't support asyncio yet - just monkey patch and c10k is OK. Oh it doesn't support 3.7? - This is lost user.

**Comment by nat-goodspeed** [comment]
Okay. Discussion so far has not been a resounding, "Go for it!!"

If we believe that this non-ideal solution will work "well enough" for now, and better than the status quo, then let's get it done. @Dormouse759, will you please merge tests/hub_test.py as indicated in the status panel?

**Comment by Dormouse759** [comment]
Here you go.
